File: src/main/java/com/jitong/im/group/GroupModels.java
package com.jitong.im.group;

import java.util.List;
import java.time.Instant;
import java.util.UUID;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;

record GroupCreateResponse(
        int version,
        UUID conversationId,
        String groupNo,
        String name,
        String description,
        String visibility,
        UUID ownerUserId,
        String role,
        String avatarUrl,
        long avatarVersion,
        int memberCount
) {
}

record GroupSummary(
        int version,
        UUID conversationId,
        String groupNo,
        String name,
        String description,
        String visibility,
        String role,
        String avatarUrl,
        long avatarVersion,
        int memberCount
) {
}

record GroupSearchResult(
        String name,
        String avatarUrl,
        String description,
        int memberCount
) {
}

record GroupSearchPage(
        int version,
        List<GroupSearchResult> groups
) {
}

record GroupMemberAddRequest(
        @Pattern(regexp = "[1-9][0-9]{10}")
        String accountNo
) {
}

record GroupMemberAddResponse(
        int version,
        UUID conversationId,
        UUID userId,
        String role,
        int memberCount
) {
}

record GroupRoleChangeRequest(
        @Pattern(regexp = "ADMIN|MEMBER")
        String role
) {
}

record GroupRoleChangeResponse(
        int version,
        UUID conversationId,
        UUID userId,
        String role
) {
}

record GroupOwnerTransferRequest(
        UUID userId
) {
}

record GroupOwnerTransferResponse(
        int version,
        UUID conversationId,
        UUID previousOwnerUserId,
        UUID ownerUserId
) {
}

record GroupProfileUpdateRequest(
        @Size(min = 1, max = 128)
        String name,
        @Size(max = 1000)
        String description,
        @Pattern(regexp = "PUBLIC|UNLISTED|PRIVATE")
        String visibility
) {
}

record GroupInviteCreateRequest(
        @Min(1)
        @Max(10000)
        Integer maxUses,
        @Min(60)
        @Max(2592000)
        Long expiresInSeconds
) {
}

record GroupInviteResponse(
        int version,
        UUID inviteId,
        UUID conversationId,
        int maxUses,
        int useCount,
        Instant expiresAt,
        String deepLink,
        String qrPayload
) {
}

record GroupInviteResolveResponse(
        int version,
        UUID conversationId,
        String groupNo,
        String name,
        String description,
        String visibility,
        String avatarUrl,
        long avatarVersion,
        int memberCount,
        Instant expiresAt
) {
}

record GroupJoinRequestCreateRequest(
        @Size(min = 20, max = 128)
        String inviteToken
) {
}

record GroupJoinRequestResponse(
        int version,
        UUID requestId,
        UUID conversationId,
        UUID userId,
        String status,
        UUID inviteId,
        Instant createdAt,
        Instant resolvedAt
) {
}

record GroupJoinRequestSummary(
        int version,
        UUID requestId,
        UUID conversationId,
        UUID userId,
        String accountNo,
        String displayName,
        String status,
        UUID inviteId,
        Instant createdAt,
        Instant resolvedAt
) {
}

record GroupBanRequest(
        @Size(max = 500)
        String reason
) {
}

File: android-app/app/src/main/java/com/jitong/im/android/group/GroupModels.kt
package com.jitong.im.android.group

import java.util.UUID

internal data class CreateGroupRequest(
    val name: String,
    val description: String,
    val visibility: String,
)

internal data class GroupCreateResponse(
    val version: Int,
    val conversationId: UUID,
    val groupNo: String,
    val name: String,
    val description: String,
    val visibility: String,
    val ownerUserId: UUID,
    val role: String,
    val avatarUrl: String?,
    val avatarVersion: Long,
    val memberCount: Int,
    val aiEnabled: Boolean = false,
    val aiPolicyVersion: Long = 1,
)

internal data class GroupSummary(
    val version: Int,
    val conversationId: UUID,
    val groupNo: String,
    val name: String,
    val description: String,
    val visibility: String,
    val role: String,
    val avatarUrl: String?,
    val avatarVersion: Long,
    val memberCount: Int,
    val aiEnabled: Boolean = false,
    val aiPolicyVersion: Long = 1,
    val unreadCount: Int = 0,
    val latestMessage: GroupPreview? = null,
)

internal data class GroupPreview(
    val conversationSeq: Long,
    val type: String,
    val text: String?,
    val state: String,
    val serverAcceptedAt: String,
    val systemEventType: String? = null,
) {
    val sortTimestamp: Long
        get() = runCatching {
            java.time.Instant.parse(serverAcceptedAt).toEpochMilli()
        }.getOrDefault(Long.MIN_VALUE)
}

internal fun GroupPreview.displayText(): String = when {
    type == "SYSTEM" && systemEventType == "GROUP_CREATED" -> "群聊已创建"
    state == "RECALLED" -> "消息已撤回"
    state == "MODERATED" -> "消息已被移除"
    type == "IMAGE" -> "图片"
    type == "TEXT" && !text.isNullOrBlank() -> text
    else -> "消息"
}

internal data class GroupSearchResult(
    val name: String,
    val groupNo: String,
    val description: String,
    val avatarUrl: String?,
    val memberCount: Int,
)

internal data class GroupSearchPage(
    val version: Int,
    val groups: List<GroupSearchResult>,
)

internal data class GroupInviteCreateRequest(
    val maxUses: Int? = null,
    val expiresInSeconds: Long? = null,
)

internal data class GroupInviteResponse(
    val version: Int,
    val inviteId: UUID,
    val conversationId: UUID,
    val maxUses: Int,
    val useCount: Int,
    val expiresAt: String,
    val deepLink: String,
    val qrPayload: String,
)

internal data class GroupInviteResolveResponse(
    val version: Int,
    val conversationId: UUID,
    val groupNo: String,
    val name: String,
    val description: String,
    val visibility: String,
    val avatarUrl: String?,
    val avatarVersion: Long,
    val memberCount: Int,
    val expiresAt: String,
)

internal data class GroupJoinRequestCreateRequest(
    val inviteToken: String? = null,
)

internal data class GroupJoinRequestByGroupNoRequest(
    val groupNo: String,
    val inviteToken: String? = null,
)

internal data class GroupJoinRequestResponse(
    val version: Int,
    val requestId: UUID,
    val conversationId: UUID,
    val userId: UUID,
    val status: String,
    val inviteId: UUID?,
    val createdAt: String,
    val resolvedAt: String?,
)

internal data class GroupJoinRequestSummary(
    val version: Int,
    val requestId: UUID,
    val conversationId: UUID,
    val userId: UUID,
    val accountNo: String,
    val displayName: String,
    val status: String,
    val inviteId: UUID?,
    val createdAt: String,
    val resolvedAt: String?,
)

internal data class GroupMemberSummary(
    val version: Int,
    val userId: UUID,
    val accountNo: String,
    val displayName: String,
    val role: String,
    val avatarUrl: String?,
    val avatarVersion: Long,
    val avatarFallback: String,
)

internal data class GroupBanRequest(
    val reason: String? = null,
)

internal data class GroupMemberInvitationRequest(
    val accountNo: String,
)

internal data class GroupMemberInvitationResponse(
    val version: Int,
    val invitationId: UUID,
    val conversationId: UUID,
    val inviterUserId: UUID,
    val inviteeUserId: UUID,
    val status: String,
    val createdAt: String,
    val resolvedAt: String?,
)

internal data class GroupMemberInvitationSummary(
    val version: Int,
    val invitationId: UUID,
    val conversationId: UUID,
    val groupNo: String,
    val groupName: String,
    val inviterUserId: UUID,
    val inviterAccountNo: String,
    val inviterDisplayName: String,
    val status: String,
    val createdAt: String,
    val resolvedAt: String?,
)

internal data class GroupRoleChangeRequest(
    val role: String,
)

internal data class GroupRoleChangeResponse(
    val version: Int,
    val conversationId: UUID,
    val userId: UUID,
    val role: String,
)

internal data class GroupOwnerTransferRequest(
    val userId: UUID,
)

internal data class GroupOwnerTransferResponse(
    val version: Int,
    val conversationId: UUID,
    val previousOwnerUserId: UUID,
    val ownerUserId: UUID,
)

internal data class GroupProfileUpdateRequest(
    val name: String,
    val description: String,
    val visibility: String,
)
